cpp/docs/DOCUMENTATION.md

```diff
@@ -0,0 +1,453 @@
+
+Doxygen supports syntax highlighting for C++ and several other programming languages (e.g. Python, Java).
+By default, the `@code` tag will use syntax highlighting based on the source code where it was found.
+
+```c++
+ *
+ * @code
+ * auto result = cudf::make_column( );
+ * @endcode
+ *
+```
+
+You can specify a different language by indicating the file extension in the tag:
+
+```c++
+ *
+ * @code{.py}
+ * import cudf
+ * s = cudf.Series([1,2,3])
+ * @endcode
+ *
+```
+
+If you wish to use pseudo-code in your example, use the following:
+
+```c++
+ *
+ * Sometimes pseudo-code is clearer.
+ * @code{.pseudo}
+ * s = int column of [ 1, 2, null, 4 ]
+ * r = fill( s, [1, 2], 0 )
+ * r is now [ 1, 0, 0, 4 ]
+ * @endcode
+ *
+```
+
+When writing example snippets, using fully qualified class names allows doxygen to add reference links to the example.
+
+```c++
+ *
+ * @code
+ * auto result1 = make_column( ); // reference link will not be created
+ * auto result2 = cudf::make_column( ); // reference link will be created
+ * @endcode
+ *
+```
+
+Although using 3 backtick marks `` ``` `` for example blocks will work too, they do not standout as well in vs-code and other source editors.
+
+Do not use the `@example` tag in the comments for a declaration.
+Doxygen will interpret the entire source file as example source code when it encounters this tag.
+The source file is then published under a separate the _Examples_ page in the output.
+
+## Namespaces
+
+Doxygen output includes a _Namespaces_ page that shows all the namespaces declared with comment blocks in the processed files.
+Here is an example of doxygen description comment for a namespace declaration.
+
+```c++
+//! cuDF interfaces
+namespace cudf {
+```
+
+A description comment should be included for only once for each unique namespace declaration.
+Otherwise, if more than one description is found, doxygen will aggregate the descriptions in an arbitrary order in the output pages.
+
+If you are introducing a new namespace, provide a short description block for only one declaration and not for every occurrence.
+
+## Groups/Modules
+
+Grouping declarations into modules helps users to find APIs in the doxygen pages.
+Generally, common functions are already grouped logically into header files but not easily reflected in the doxygen output.
+Doxygen output includes a _Modules_ page that organizes items into groups using the [Grouping doxygen commands](http://www.doxygen.nl/manual/grouping.html)
+These commands can group common functions across header files, source files, and even namespaces.
+Groups can also be nested by defining new groups within existing groups.
+
+For libcudf, all the group hierarchy is defined in the `doxygen_groups.h` header file.
+This file does not need to be included in any other source file.
+It is processed only by the doxygen tool.
+Use this file to add or update any groups.
+Doxygen will also use the definitions in this file to order to groups in the _Modules_ page.
```

Comment:
```suggestion
`doxygen_groups.h` does not need to be included in any other source file, because the definitions in this file are used only by the doxygen tool to generate groups in the _Modules_ page. Modify this file only to add or update groups. The existing groups have been carefully structured and named, so new groups should be added thoughtfully.
```